sdk/core/azure-core/inc/azure/core/internal/deferred_operation.hpp

```diff
@@ -0,0 +1,173 @@
+     *
+     * Capturing the parameters to the batched operation also avoids a potential issue
+     * associated with the lifetime of authentication tokens - if the lifetime of the
+     * authentication token is short (15 minutes or so), it is possible that the token
+     * will expire between when the DeferredResponse object is created and when the
+     * batched operation is submitted to the server.
+     */
+    class DeferredResponseFactory {
+      std::vector<std::shared_ptr<DeferredResponseSharedBase>> m_deferredOperations;
+
+    protected:
+      virtual std::vector<std::shared_ptr<DeferredResponseSharedBase>> const& DeferredOperations()
+      {
+        return m_deferredOperations;
+      }
+
+    public:
+      /***
+       * @brief Creates a deferred operation from a customer supplied shared object derived from
+       * DeferredResponseSharedBase.
+       *
+       * This overload allows a customer to derive their own DeferredOperationSharedBase derived
+       * class which allows them to process deferred operations without providing a lambda (thus
+       * avoiding the challenges associated with lambda captures).
+       *
+       */
+      template <typename T>
+      DeferredResponse<T> CreateDeferredResponse(
+          std::shared_ptr<DeferredResponseSharedBase> deferredOperationShared)
+      {
+        m_deferredOperations.push_back(deferredOperationShared);
+        return DeferredResponse<T>(deferredOperationShared);
+      }
+
+      /***
+       * @brief Creates a deferred operation from the supplied HTTP request object.
+       *
+       * Creates a deferred operation with a callback function which can be used by the caller to
+       * process any results returned by the service.
+       *
+       * NOTE: If the callback function is a C++ lambda, the lambda cannot capture any values by
+       * reference - it absolutely will be called in a context outside the function in which the
+       * lambda was created.
+       *
+       */
+      template <typename T>
+      DeferredResponse<T> CreateDeferredResponse(
+          Azure::Core::Http::Request requestToDefer,
+          std::function<Azure::Response<T>(std::unique_ptr<Azure::Core::Http::RawResponse>&)>
+              completeProcessing)
+      {
+        std::shared_ptr<DeferredResponseSharedBase> deferredResponse(
+            std::make_shared<DeferredResponseShared<T>>(requestToDefer, completeProcessing));
+        m_deferredOperations.push_back(deferredResponse);
+        return DeferredResponse<T>(deferredResponse);
+      }
+    };
+
+  } // namespace _internal
+
+  template <typename T> class DeferredResponse {
+  private:
+    std::shared_ptr<_internal::DeferredResponseSharedBase> m_sharedState;
+
+    friend class _internal::DeferredResponseFactory;
+    DeferredResponse(std::shared_ptr<_internal::DeferredResponseSharedBase> sharedState)
+        : m_sharedState(sharedState)
+    {
+    }
+
+  public:
+    Response<T> GetResponse() const
```

Comment:
What is expected behavior if the customer calls this function before submitting batch?

Reply:
"The results are undefined." 

It's completely reasonable for the code to throw an exception. It's also completely reasonable that we access violate.

The same is true if GetResponse() is called twice - the results are undefined. 

Personally I'd throw an exception, but it's not a requirement of the pattern.

The guidelines for the Azure SDK are that we don't normally check for precondition failures.